api/v1alpha1/httproute_types.go

```diff
@@ -179,20 +224,101 @@ type HTTPHeaderFilter struct {
 type HTTPRouteAction struct {
 	// ForwardTo sends requests to the referenced object.
-	ForwardTo *core.TypedLocalObjectReference `json:"forwardTo"`
+	ForwardTo *RouteActionTargetObjectReference `json:"forwardTo"`
 
 	// Extension is an optional, implementation-specific extension
 	// to the "action" behavior.
 	//
 	// Support: custom
 	//
 	// +optional
-	Extension *core.TypedLocalObjectReference `json:"extension"`
+	Extension *RouteActionExtensionObjectReference `json:"extension"`
+}
+
+// RouteActionTargetObjectReference identifies a target object for a route
+// action within a known namespace.
+type RouteActionTargetObjectReference struct {
+	// Group is the group of the referent.  For example,
+	// "networking.acme.io".  The empty string represents the core API
+	// group.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Group string `json:"group,omitempty"`
+	// Resource is the resource of the referent.  For example, "service" or
+	// "myroutetarget".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Resource string `json:"resource"`
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+}
+
+// RouteActionExtensionObjectReference identifies a route-action extension
+// object within a known namespace.
+type RouteActionExtensionObjectReference struct {
+	// Group is the group of the referent.  For example,
+	// "networking.acme.io".  The empty string represents the core API
+	// group.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Group string `json:"group,omitempty"`
+	// Resource is the resource of the referent.  For example, "configmap"
+	// or "myrouteaction".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Resource string `json:"resource"`
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
 }
 
-// HTTPRouteStatus defines the observed state of HTTPRoute
+// RouteHostExtensionObjectReference identifies a route-host extension
+// object within a known namespace.
+type RouteHostExtensionObjectReference struct {
+	// Group is the group of the referent.  For example,
+	// "networking.acme.io".  The empty string represents the core API
+	// group.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Group string `json:"group,omitempty"`
+	// Resource is the resource of the referent.  For example, "configmap"
+	// or "myroutehost".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Resource string `json:"resource"`
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+}
+
+// HTTPRouteStatus defines the observed state of HTTPRoute.
 type HTTPRouteStatus struct {
-	Gateways []core.ObjectReference `json:"gateways"`
+	Gateways []GatewayObjectReference `json:"gateways"`
+}
+
+// GatewayObjectReference identifies a Gateway object.
+type GatewayObjectReference struct {
```

Comment:
Should this definition be in `gateway_types.go`?

Reply:
Yes, will be used by more than http imagine

Reply:
Put that in a follow on PR.